File: packages/treeBurst/src/treeBurst/runtime/ScriptFunction.ts
import { Expression } from "../syntax/Expression"
import { evaluateExpression, LABEL_RETURN } from "./evaluateExpression"
import { ExpressionResult } from "./ExpressionResult"
import { VOID } from "./GlobalScope"
import { ManagedFunction } from "./ManagedFunction"
import { ManagedObject } from "./ManagedObject"
import { ManagedValue } from "./ManagedValue"
import { Scope } from "./Scope"

export class ScriptFunction extends ManagedFunction {
    public override invoke(args: ManagedValue[], scope: Scope, result: ExpressionResult): void {
        const functionScope = scope.makeChild()

        for (let i = 0; i < this.parameters.length; i++) {
            const parameter = this.parameters[i]
            const arg = i < args.length ? args[i] : VOID

            functionScope.declareVariable(parameter)!.value = arg
        }

        evaluateExpression(this.body, functionScope, result)
        if (result.label == LABEL_RETURN) {
            result.label = null
        }

        return
    }

    constructor(
        prototype: ManagedObject | null,
        parameters: string[],
        public readonly body: Expression,
        public readonly scope: Scope,
    ) { super(prototype, parameters) }
}

File: packages/treeBurstJava/src/main/java/bt7s7k7/treeburst/runtime/ScriptFunction.java
package bt7s7k7.treeburst.runtime;

import static bt7s7k7.treeburst.runtime.ExpressionResult.LABEL_EXCEPTION;
import static bt7s7k7.treeburst.runtime.ExpressionResult.LABEL_RETURN;

import java.util.List;

import bt7s7k7.treeburst.bytecode.ProgramFragment;
import bt7s7k7.treeburst.support.Diagnostic;
import bt7s7k7.treeburst.support.ManagedValue;
import bt7s7k7.treeburst.support.Parameter;
import bt7s7k7.treeburst.support.Position;

public class ScriptFunction extends ManagedFunction {
	public final ProgramFragment body;
	public final Scope scope;
	public List<Parameter> parameterDeclarations;

	@Override
	public void invoke(List<ManagedValue> args, Scope scope, ExpressionResult result) {
		Scope functionScope = this.scope.makeChild();

		Parameter.destructure(this.getParameters(), true, args, functionScope, result);
		if (result.label != null) return;

		this.body.evaluate(functionScope, result);
		if (result.label == null) return;

		if (LABEL_RETURN.equals(result.label)) {
			result.label = null;
			return;
		}

		if (LABEL_EXCEPTION.equals(result.label)) {
			return;
		}

		var oldLabel = result.label;
		result.label = null;
		result.setException(new Diagnostic("Did not resolve label '" + oldLabel + "' during function execution", Position.INTRINSIC));
	}

	public ScriptFunction(ManagedObject prototype, List<Parameter> parameters, ProgramFragment body, Scope scope) {
		super(prototype, null, parameters);
		this.body = body;
		this.scope = scope;
	}
}
